src/Components/Web.JS/src/Services/NavigationManager.ts

```diff
@@ -87,6 +105,23 @@ export function attachToEventDelegator(eventDelegator: EventDelegator): void {
   });
 }
 
+function isSamePageWithHash(absoluteHref: string): boolean {
+  const hashIndex = absoluteHref.indexOf('#');
+  return hashIndex > -1 && location.href.replace(location.hash, '') === absoluteHref.substring(0, hashIndex);
+}
+
+function performScrollToElementOnTheSamePage(absoluteHref : string): void {
+  history.pushState({}, "", absoluteHref);
```

Comment:
In the [one other place](https://github.com/dotnet/aspnetcore/blob/90efca2f6a7682e0c952f26dfe1147c7766cd9d2/src/Components/Web.JS/src/Services/NavigationManager.ts#L199) we already call `pushState`, we include an `_index` property in the object that we pass as the `state` parameter. Could we do that here as well? Feel free to make a helper function to reduce code duplication.

---

Just for some context (and this really isn't that relevant to your PR so feel free to ignore, but I'm writing it for completeness), the reason we store a history "index" is to support an edge case in the "navigation cancellation" feature. When enabled, that feature has the requirement that we need to be able to "revert" navigations as soon as they happen, then get asynchronous confirmation from the .NET side to either replay or abort the navigation.

The edge case is when the user's navigation skips over multiple history entries (e.g., via the right-click menu for the browser's forward/back buttons). We need a way of detecting how far back/forward the navigation went so we know how far to move through the history stack to revert the navigation. An "index" in each history entry allows us to do the math required to perform that operation.

Without persisting the index here, I _think_ a bug would occur when the "location changing" event is enabled and the user navigates back to the history entry caused by the scroll navigation. An undefined `_index` is interpreted as an index of `0`, which could send the user way further back in history than they intended.

Reply:
Thanks for ringing that up!